libs/x-content/src/main/java/org/elasticsearch/common/xcontent/MediaTypeParser.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.xcontent;
+
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+
+public class MediaTypeParser<T extends MediaType> {
+    private final Map<String, T> formatToMediaType = new HashMap<>();
+    private final Map<String, T> typeWithSubtypeToMediaType = new HashMap<>();
+
+    public MediaTypeParser(T[] acceptedMediaTypes) {
+        for (T mediaType : acceptedMediaTypes) {
+            typeWithSubtypeToMediaType.put(mediaType.typeWithSubtype(), mediaType);
+            formatToMediaType.put(mediaType.format(), mediaType);
+        }
+    }
+
+    public MediaTypeParser(T[] acceptedMediaTypes, Map<String, T> additionalMediaTypes) {
+        this(acceptedMediaTypes);
+        for (Map.Entry<String, T> entry : additionalMediaTypes.entrySet()) {
+            String typeWithSubtype = entry.getKey();
+            T mediaType = entry.getValue();
+
+            typeWithSubtypeToMediaType.put(typeWithSubtype.toLowerCase(Locale.ROOT), mediaType);
+            formatToMediaType.put(mediaType.format(), mediaType);
+        }
+    }
```

Comment:
This could be considered a nit and is my preference on style and immutability.

```suggestion
    private final Map<String, T> formatToMediaType;
    private final Map<String, T> typeWithSubtypeToMediaType;

    public MediaTypeParser(T[] acceptedMediaTypes) {
        this(acceptedMediaTypes, Map.of());
    }

    public MediaTypeParser(T[] acceptedMediaTypes, Map<String, T> additionalMediaTypes) {
        final int size = acceptedMediaTypes.length + additionalMediaTypes.size();
        Map<String, T> formatMap = new HashMap<>(size);
        Map<String, T> typeMap = new HashMap<>(size);
        for (T mediaType : acceptedMediaTypes) {
            typeMap.put(mediaType.typeWithSubtype(), mediaType);
            formatMap.put(mediaType.format(), mediaType);
        }
        for (Map.Entry<String, T> entry : additionalMediaTypes.entrySet()) {
            String typeWithSubtype = entry.getKey();
            T mediaType = entry.getValue();

            typeMap.put(typeWithSubtype.toLowerCase(Locale.ROOT), mediaType);
            formatMap.put(mediaType.format(), mediaType);
        }

        this.formatToMediaType = Map.copyOf(formatMap);
        this.typeWithSubtypeToMediaType = Map.copyOf(typeMap);
    }
```